internal/metrics/otel.go

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) F5, Inc.
+//
+// This source code is licensed under the Apache License, Version 2.0 license found in the
+// LICENSE file in the root directory of this source tree.
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/nginx/agent/v3/internal/config"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials/insecure"
+)
+
+const (
+	defaultgRPCConnTimeout    = 10 * time.Second
+	defaultgRPCMinConnTimeout = 5 * time.Second
+	defaultBackoffDelay       = 240 * time.Second
+)
+
+// NewGRPCExporter returns a OTel exporter that transmits via gRPC.
+func NewGRPCExporter(ctx context.Context, conf config.Metrics) (*otlpmetricgrpc.Exporter, error) {
+	ctx, cancel := context.WithTimeout(ctx, defaultgRPCConnTimeout)
+	defer cancel()
+	// Exponential back-off strategy.
+	backoffConf := backoff.DefaultConfig
+	// You can also change the base delay, multiplier, and jitter here.
+	backoffConf.MaxDelay = defaultBackoffDelay
+
+	conn, err := grpc.DialContext(
+		ctx,
+		conf.OTelExporterTarget,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithConnectParams(grpc.ConnectParams{
+			Backoff: backoffConf,
+			// Connection timeout.
+			MinConnectTimeout: defaultgRPCMinConnTimeout,
+		}),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create grpc connection: %w", err)
+	}
+
+	return otlpmetricgrpc.New(
+		ctx, otlpmetricgrpc.WithGRPCConn(conn), otlpmetricgrpc.WithTimeout(defaultgRPCConnTimeout),
+	)
+}
+
+// NewHTTPExporter returns a OTel exporter that transmits via HTTP.
+// nolint: gomnd
+func NewHTTPExporter(ctx context.Context) (*otlpmetrichttp.Exporter, error) {
+	target := "0.0.0.0:4317"
+	if value, ok := os.LookupEnv("HTTP_TARGET"); ok {
+		target = value
+	}
+
+	return otlpmetrichttp.New(
+		ctx,
+		otlpmetrichttp.WithInsecure(),
+		otlpmetrichttp.WithEndpoint(target),
+		// WithTimeout sets the max amount of time the Exporter will attempt an
+		// export.
+		otlpmetrichttp.WithTimeout(7*time.Second),
+		otlpmetrichttp.WithRetry(otlpmetrichttp.RetryConfig{
```

Comment:
we probably should externalize these configurations

Reply:
When you say "externalize", do you mean read them from our config file?

Reply:
The gRPC configurations are now taken from the Agent's config file, but the HTTP ones are not, as we don't currently support HTTP OTel connections.

Reply:
The HTTPExporter was removed per: https://github.com/nginx/agent/pull/544#discussion_r1497688634